src/table.js

```diff
@@ -445,64 +450,168 @@ Table.prototype.createReadStream = function(options) {
+    var grpcOpts = {
+      service: 'Bigtable',
+      method: 'readRows',
+      retryOpts: {
+        currentRetryAttempt: numRequestsMade,
+      },
+    };
+
+    var reqOpts = {
+      tableName: this.id,
+      objectMode: true,
+    };
+    if (lastRowKey) {
+      const lessThan = (lhs, rhs) => {
+        return (
+          Mutation.convertToBytes(lhs).compare(Mutation.convertToBytes(rhs)) ===
+          -1
+        );
+      };
+      const greaterThan = (lhs, rhs) => {
+        return lessThan(rhs, lhs);
+      };
+      const greaterThanOrEqualTo = (lhs, rhs) => {
+        return !lessThan(rhs, lhs);
+      };
+
+      if (ranges.length === 0) {
+        ranges.push({
+          start: {
+            value: lastRowKey,
+            inclusive: false,
+          },
+        });
+      } else {
+        // Readjust and/or remove ranges based on previous valid row reads.
+
+        // Iterate backward since items may need to be removed.
+        for (let index = ranges.length - 1; index >= 0; index--) {
+          const range = ranges[index];
+          const normalizeRangeValue = rangeSide => {
+            return is.object(rangeSide) ? rangeSide.value : rangeSide;
+          };
+          const startValue = normalizeRangeValue(range.start);
+          const endValue = normalizeRangeValue(range.end);
+          const isWithinStart =
+            !startValue || greaterThanOrEqualTo(startValue, lastRowKey);
+          const isWithinEnd = !endValue || lessThan(lastRowKey, endValue);
+          if (isWithinStart) {
+            if (isWithinEnd) {
+              // The lastRowKey is within this range, adjust the start value.
+              range.start = {
+                value: lastRowKey,
+                inclusive: false,
+              };
+            } else {
+              // The lastRowKey is past this range, remove this range.
+              ranges.splice(index, 1);
+            }
+          }
+        }
       }
-      const row = self.row(rowData.key);
-      row.data = rowData.data;
-      next(null, row);
-    }),
-  ]);
-  return stream;
+
+      // Remove rowKeys already read.
+      if (rowKeys) {
+        rowKeys = rowKeys.filter(rowKey => greaterThan(rowKey, lastRowKey));
+        if (rowKeys.length === 0) {
+          rowKeys = null;
+        }
+      }
+    }
+    if (rowKeys || ranges.length) {
+      reqOpts.rows = {};
+      if (rowKeys) {
+        reqOpts.rows.rowKeys = rowKeys.map(Mutation.convertToBytes);
+      }
+      if (ranges.length) {
+        reqOpts.rows.rowRanges = ranges.map(function(range) {
+          return Filter.createRange(range.start, range.end, 'Key');
+        });
+      }
+    }
+    if (filter) {
+      reqOpts.filter = filter;
+    }
+
+    if (rowsLimit) {
+      reqOpts.rowsLimit = rowsLimit - rowsRead;
+    }
+
+    const requestStream = this.requestStream(grpcOpts, reqOpts);
+    requestStream.on('request', () => numRequestsMade++);
+
+    const stream = pumpify.obj([
```

Comment:
Let's call this `requestStream` or `rowStream`. Ideally "requestStream" is piped to "userStream".